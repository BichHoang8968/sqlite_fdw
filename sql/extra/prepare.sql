-- Regression tests for prepareable statements. We query the content
-- of the pg_prepared_statements view as prepared statements are
-- created and removed.
CREATE EXTENSION sqlite_fdw;
CREATE SERVER sqlite_svr FOREIGN DATA WRAPPER sqlite_fdw
OPTIONS (database '/tmp/sqlitefdw_test_core.db');

CREATE FOREIGN TABLE tenk1 (
	unique1		int4,
	unique2		int4,
	two			int4,
	four		int4,
	ten			int4,
	twenty		int4,
	hundred		int4,
	thousand	int4,
	twothousand	int4,
	fivethous	int4,
	tenthous	int4,
	odd			int4,
	even		int4,
	stringu1	name,
	stringu2	name,
	string4		name
) SERVER sqlite_svr;

ALTER TABLE tenk1 SET WITH OIDS;

CREATE FOREIGN TABLE road (
	name		text,
	thepath 	path
) SERVER sqlite_svr;

--Testcase 1:
SELECT name, statement, parameter_types FROM pg_prepared_statements;

--Testcase 2:
PREPARE q1 AS SELECT * FROM road LIMIT 1;
--Testcase 3:
EXECUTE q1;

--Testcase 4:
SELECT name, statement, parameter_types FROM pg_prepared_statements;

-- should fail
--Testcase 5:
PREPARE q1 AS SELECT * FROM tenk1 LIMIT 1;

-- should succeed
DEALLOCATE q1;
--Testcase 6:
PREPARE q1 AS SELECT * FROM tenk1 LIMIT 1;
--Testcase 7:
EXECUTE q1;

--Testcase 8:
PREPARE q2 AS SELECT * FROM tenk1 LIMIT 1;
--Testcase 9:
SELECT name, statement, parameter_types FROM pg_prepared_statements;

-- sql92 syntax
DEALLOCATE PREPARE q1;

--Testcase 10:
SELECT name, statement, parameter_types FROM pg_prepared_statements;

DEALLOCATE PREPARE q2;
-- the view should return the empty set again
--Testcase 11:
SELECT name, statement, parameter_types FROM pg_prepared_statements;

-- parameterized queries
--Testcase 12:
PREPARE q2(text) AS
	SELECT datname, datistemplate, datallowconn
	FROM pg_database WHERE datname = $1;

--Testcase 13:
EXECUTE q2('postgres');

--Testcase 14:
PREPARE q3(text, int, float, boolean, oid, smallint) AS
	SELECT * FROM tenk1 WHERE string4 = $1 AND (four = $2 OR
	ten = $3::bigint OR true = $4 OR oid = $5 OR odd = $6::int)
	ORDER BY unique1;

--Testcase 15:
EXECUTE q3('AAAAxx', 5::smallint, 10.5::float, false, 500::oid, 4::bigint);

-- too few params
--Testcase 16:
EXECUTE q3('bool');

-- too many params
--Testcase 17:
EXECUTE q3('bytea', 5::smallint, 10.5::float, false, 500::oid, 4::bigint, true);

-- wrong param types
--Testcase 18:
EXECUTE q3(5::smallint, 10.5::float, false, 500::oid, 4::bigint, 'bytea');

-- invalid type
--Testcase 19:
PREPARE q4(nonexistenttype) AS SELECT * FROM road WHERE name = $1;

-- create table as execute
--Testcase 20:
PREPARE q5(int, text) AS
	SELECT * FROM tenk1 WHERE unique1 = $1 OR stringu1 = $2
	ORDER BY unique1;
CREATE TEMPORARY TABLE q5_prep_results AS EXECUTE q5(200, 'DTAAAA');
--Testcase 21:
SELECT * FROM q5_prep_results;

-- unknown or unspecified parameter types: should succeed
--Testcase 22:
PREPARE q6 AS
    SELECT * FROM tenk1 WHERE unique1 = $1 AND stringu1 = $2;
--Testcase 23:
PREPARE q7(unknown) AS
    SELECT * FROM road WHERE thepath = $1;

--Testcase 24:
SELECT name, statement, parameter_types FROM pg_prepared_statements
    ORDER BY name;

-- test DEALLOCATE ALL;
DEALLOCATE ALL;
--Testcase 25:
SELECT name, statement, parameter_types FROM pg_prepared_statements
    ORDER BY name;

DROP FOREIGN TABLE tenk1;
DROP FOREIGN TABLE road;
DROP SERVER sqlite_svr;
DROP EXTENSION sqlite_fdw CASCADE;