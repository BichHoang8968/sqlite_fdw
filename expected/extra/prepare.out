-- Regression tests for prepareable statements. We query the content
-- of the pg_prepared_statements view as prepared statements are
-- created and removed.
CREATE EXTENSION sqlite_fdw;
CREATE SERVER sqlite_svr FOREIGN DATA WRAPPER sqlite_fdw
OPTIONS (database '/tmp/sqlitefdw_test_core.db');
CREATE FOREIGN TABLE tenk1 (
	unique1		int4,
	unique2		int4,
	two			int4,
	four		int4,
	ten			int4,
	twenty		int4,
	hundred		int4,
	thousand	int4,
	twothousand	int4,
	fivethous	int4,
	tenthous	int4,
	odd			int4,
	even		int4,
	stringu1	name,
	stringu2	name,
	string4		name
) SERVER sqlite_svr;
ALTER TABLE tenk1 SET WITH OIDS;
ERROR:  syntax error at or near "WITH"
LINE 1: ALTER TABLE tenk1 SET WITH OIDS;
                              ^
CREATE FOREIGN TABLE road (
	name		text,
	thepath 	path
) SERVER sqlite_svr;
--Testcase 1:
SELECT name, statement, parameter_types FROM pg_prepared_statements;
 name | statement | parameter_types 
------+-----------+-----------------
(0 rows)

--Testcase 2:
PREPARE q1 AS SELECT * FROM road LIMIT 1;
--Testcase 3:
EXECUTE q1;
                name                |                 thepath                 
------------------------------------+-----------------------------------------
 A                             St   | [(-122.0265,37.049),(-122.0271,37.045)]
(1 row)

--Testcase 4:
SELECT name, statement, parameter_types FROM pg_prepared_statements;
 name |                 statement                 | parameter_types 
------+-------------------------------------------+-----------------
 q1   | PREPARE q1 AS SELECT * FROM road LIMIT 1; | {}
(1 row)

-- should fail
--Testcase 5:
PREPARE q1 AS SELECT * FROM tenk1 LIMIT 1;
ERROR:  prepared statement "q1" already exists
-- should succeed
DEALLOCATE q1;
--Testcase 6:
PREPARE q1 AS SELECT * FROM tenk1 LIMIT 1;
--Testcase 7:
EXECUTE q1;
 unique1 | unique2 | two | four | ten | twenty | hundred | thousand | twothousand | fivethous | tenthous | odd | even | stringu1 | stringu2 | string4 
---------+---------+-----+------+-----+--------+---------+----------+-------------+-----------+----------+-----+------+----------+----------+---------
    8800 |       0 |   0 |    0 |   0 |      0 |       0 |      800 |         800 |      3800 |     8800 |   0 |    1 | MAAAAA   | AAAAAA   | AAAAxx
(1 row)

--Testcase 8:
PREPARE q2 AS SELECT * FROM tenk1 LIMIT 1;
--Testcase 9:
SELECT name, statement, parameter_types FROM pg_prepared_statements;
 name |                 statement                  | parameter_types 
------+--------------------------------------------+-----------------
 q1   | PREPARE q1 AS SELECT * FROM tenk1 LIMIT 1; | {}
 q2   | PREPARE q2 AS SELECT * FROM tenk1 LIMIT 1; | {}
(2 rows)

-- sql92 syntax
DEALLOCATE PREPARE q1;
--Testcase 10:
SELECT name, statement, parameter_types FROM pg_prepared_statements;
 name |                 statement                  | parameter_types 
------+--------------------------------------------+-----------------
 q2   | PREPARE q2 AS SELECT * FROM tenk1 LIMIT 1; | {}
(1 row)

DEALLOCATE PREPARE q2;
-- the view should return the empty set again
--Testcase 11:
SELECT name, statement, parameter_types FROM pg_prepared_statements;
 name | statement | parameter_types 
------+-----------+-----------------
(0 rows)

-- parameterized queries
--Testcase 12:
PREPARE q2(text) AS
	SELECT datname, datistemplate, datallowconn
	FROM pg_database WHERE datname = $1;
--Testcase 13:
EXECUTE q2('postgres');
 datname  | datistemplate | datallowconn 
----------+---------------+--------------
 postgres | f             | t
(1 row)

--Testcase 14:
PREPARE q3(text, int, float, boolean, oid, smallint) AS
	SELECT * FROM tenk1 WHERE string4 = $1 AND (four = $2 OR
	ten = $3::bigint OR true = $4 OR oid = $5 OR odd = $6::int)
	ORDER BY unique1;
ERROR:  column "oid" does not exist
LINE 3:  ten = $3::bigint OR true = $4 OR oid = $5 OR odd = $6::int)
                                          ^
HINT:  Perhaps you meant to reference the column "tenk1.odd".
--Testcase 15:
EXECUTE q3('AAAAxx', 5::smallint, 10.5::float, false, 500::oid, 4::bigint);
ERROR:  prepared statement "q3" does not exist
-- too few params
--Testcase 16:
EXECUTE q3('bool');
ERROR:  prepared statement "q3" does not exist
-- too many params
--Testcase 17:
EXECUTE q3('bytea', 5::smallint, 10.5::float, false, 500::oid, 4::bigint, true);
ERROR:  prepared statement "q3" does not exist
-- wrong param types
--Testcase 18:
EXECUTE q3(5::smallint, 10.5::float, false, 500::oid, 4::bigint, 'bytea');
ERROR:  prepared statement "q3" does not exist
-- invalid type
--Testcase 19:
PREPARE q4(nonexistenttype) AS SELECT * FROM road WHERE name = $1;
ERROR:  type "nonexistenttype" does not exist
LINE 1: PREPARE q4(nonexistenttype) AS SELECT * FROM road WHERE name...
                   ^
-- create table as execute
--Testcase 20:
PREPARE q5(int, text) AS
	SELECT * FROM tenk1 WHERE unique1 = $1 OR stringu1 = $2
	ORDER BY unique1;
CREATE TEMPORARY TABLE q5_prep_results AS EXECUTE q5(200, 'DTAAAA');
--Testcase 21:
SELECT * FROM q5_prep_results;
 unique1 | unique2 | two | four | ten | twenty | hundred | thousand | twothousand | fivethous | tenthous | odd | even | stringu1 | stringu2 | string4 
---------+---------+-----+------+-----+--------+---------+----------+-------------+-----------+----------+-----+------+----------+----------+---------
     200 |    9441 |   0 |    0 |   0 |      0 |       0 |      200 |         200 |       200 |      200 |   0 |    1 | SHAAAA   | DZNAAA   | HHHHxx
     497 |    9092 |   1 |    1 |   7 |     17 |      97 |      497 |         497 |       497 |      497 | 194 |  195 | DTAAAA   | SLNAAA   | AAAAxx
    1173 |    6699 |   1 |    1 |   3 |     13 |      73 |      173 |        1173 |      1173 |     1173 | 146 |  147 | DTAAAA   | RXJAAA   | VVVVxx
    1849 |    8143 |   1 |    1 |   9 |      9 |      49 |      849 |        1849 |      1849 |     1849 |  98 |   99 | DTAAAA   | FBMAAA   | VVVVxx
    2525 |      64 |   1 |    1 |   5 |      5 |      25 |      525 |         525 |      2525 |     2525 |  50 |   51 | DTAAAA   | MCAAAA   | AAAAxx
    3201 |    7309 |   1 |    1 |   1 |      1 |       1 |      201 |        1201 |      3201 |     3201 |   2 |    3 | DTAAAA   | DVKAAA   | HHHHxx
    3877 |    4060 |   1 |    1 |   7 |     17 |      77 |      877 |        1877 |      3877 |     3877 | 154 |  155 | DTAAAA   | EAGAAA   | AAAAxx
    4553 |    4113 |   1 |    1 |   3 |     13 |      53 |      553 |         553 |      4553 |     4553 | 106 |  107 | DTAAAA   | FCGAAA   | HHHHxx
    5229 |    6407 |   1 |    1 |   9 |      9 |      29 |      229 |        1229 |       229 |     5229 |  58 |   59 | DTAAAA   | LMJAAA   | VVVVxx
    5905 |    9537 |   1 |    1 |   5 |      5 |       5 |      905 |        1905 |       905 |     5905 |  10 |   11 | DTAAAA   | VCOAAA   | HHHHxx
    6581 |    4686 |   1 |    1 |   1 |      1 |      81 |      581 |         581 |      1581 |     6581 | 162 |  163 | DTAAAA   | GYGAAA   | OOOOxx
    7257 |    1895 |   1 |    1 |   7 |     17 |      57 |      257 |        1257 |      2257 |     7257 | 114 |  115 | DTAAAA   | XUCAAA   | VVVVxx
    7933 |    4514 |   1 |    1 |   3 |     13 |      33 |      933 |        1933 |      2933 |     7933 |  66 |   67 | DTAAAA   | QRGAAA   | OOOOxx
    8609 |    5918 |   1 |    1 |   9 |      9 |       9 |      609 |         609 |      3609 |     8609 |  18 |   19 | DTAAAA   | QTIAAA   | OOOOxx
    9285 |    8469 |   1 |    1 |   5 |      5 |      85 |      285 |        1285 |      4285 |     9285 | 170 |  171 | DTAAAA   | TNMAAA   | HHHHxx
    9961 |    2058 |   1 |    1 |   1 |      1 |      61 |      961 |        1961 |      4961 |     9961 | 122 |  123 | DTAAAA   | EBDAAA   | OOOOxx
(16 rows)

-- unknown or unspecified parameter types: should succeed
--Testcase 22:
PREPARE q6 AS
    SELECT * FROM tenk1 WHERE unique1 = $1 AND stringu1 = $2;
--Testcase 23:
PREPARE q7(unknown) AS
    SELECT * FROM road WHERE thepath = $1;
--Testcase 24:
SELECT name, statement, parameter_types FROM pg_prepared_statements
    ORDER BY name;
 name |                            statement                            | parameter_types 
------+-----------------------------------------------------------------+-----------------
 q2   | PREPARE q2(text) AS                                            +| {text}
      |         SELECT datname, datistemplate, datallowconn            +| 
      |         FROM pg_database WHERE datname = $1;                    | 
 q5   | PREPARE q5(int, text) AS                                       +| {integer,text}
      |         SELECT * FROM tenk1 WHERE unique1 = $1 OR stringu1 = $2+| 
      |         ORDER BY unique1;                                       | 
 q6   | PREPARE q6 AS                                                  +| {integer,name}
      |     SELECT * FROM tenk1 WHERE unique1 = $1 AND stringu1 = $2;   | 
 q7   | PREPARE q7(unknown) AS                                         +| {path}
      |     SELECT * FROM road WHERE thepath = $1;                      | 
(4 rows)

-- test DEALLOCATE ALL;
DEALLOCATE ALL;
--Testcase 25:
SELECT name, statement, parameter_types FROM pg_prepared_statements
    ORDER BY name;
 name | statement | parameter_types 
------+-----------+-----------------
(0 rows)

DROP FOREIGN TABLE tenk1;
DROP FOREIGN TABLE road;
DROP SERVER sqlite_svr;
DROP EXTENSION sqlite_fdw CASCADE;
